File: src/main/java/com/kasisoft/libs/common/pools/Bucket.java
package com.kasisoft.libs.common.pools;

import jakarta.validation.constraints.*;

import java.util.function.*;

import java.util.*;

import java.lang.ref.*;

/**
 * Collector for often used objects like collections, maps etc. .
 *
 * @author daniel.kasmeroglu@kasisoft.net
 */
public class Bucket<T> {

    private List<SoftReference<T>> references;

    private Supplier<T>            creator;

    private Consumer<T>            reset;

    /**
     * Initializes this bucket.
     *
     * @param producer
     *            Supplier for new elements.
     * @param resetter
     *            Cleaning function for elements.
     */
    public Bucket(@NotNull Supplier<T> producer, @NotNull Consumer<T> resetter) {
        references = new LinkedList<>();
        creator    = producer;
        reset      = resetter;
    }

    public void reset() {
        synchronized (references) {
            references.forEach(SoftReference::clear);
            references.clear();
        }
    }

    /**
     * Returns the number of references currently stored.
     *
     * @return The number of references currently stored.
     */
    public int getSize() {
        return references.size();
    }

    /**
     * Creates a new object (potentially a reused one).
     *
     * @return A new object.
     */
    public @NotNull T allocate() {
        T result = null;
        synchronized (references) {
            while ((result == null) && (!references.isEmpty())) {
                var reference = references.remove(0);
                result = reference.get();
                reference.clear();
            }
        }
        if (result == null) {
            result = creator.get();
        }
        return result;
    }

    /**
     * Frees the supplied object, so it's allowed to be reused.
     *
     * @param object
     *            The object that shall be freed.
     */
    public <R extends T> void free(R object) {
        if (object != null) {
            synchronized (references) {
                reset.accept(object);
                references.add(new SoftReference<>(object));
            }
        }
    }

    /**
     * Executes the supplied function with the desired instance.
     *
     * @param function
     *            The function that is supposed to be executed.
     * @return The return value of the supplied function.
     */
    public <R> R forInstance(@NotNull Function<T, R> function) {
        T instance = allocate();
        try {
            return function.apply(instance);
        } finally {
            free(instance);
        }
    }

    /**
     * Executes the supplied function with the desired instance.
     *
     * @param function
     *            The function that is supposed to be executed.
     * @param param
     *            An additional parameter for the function.
     * @return The return value of the supplied function.
     */
    public <R, P> R forInstance(@NotNull BiFunction<T, P, R> function, P param) {
        T instance = allocate();
        try {
            return function.apply(instance, param);
        } finally {
            free(instance);
        }
    }

    /**
     * Executes the supplied consumer with the desired instance.
     *
     * @param consumer
     *            The consumer that is supposed to be executed.
     */
    public void forInstanceDo(@NotNull Consumer<T> consumer) {
        forInstance($ -> {
            consumer.accept($);
            return null;
        });
    }

    /**
     * Executes the supplied consumer with the desired instance.
     *
     * @param consumer
     *            The consumer that is supposed to be executed.
     * @param param
     *            An additional parameter for the function.
     */
    public <P> void forInstanceDo(@NotNull BiConsumer<T, P> consumer, P param) {
        forInstance($ -> {
            consumer.accept($, param);
            return null;
        });
    }

} /* ENDCLASS */

File: src/main/kotlin/com/kasisoft/libs/common/pools/Bucket.kt
package com.kasisoft.libs.common.pools

import java.util.function.Function
import java.util.function.Consumer
import java.util.function.Supplier
import java.util.*

import java.lang.ref.*

/**
 * Collector for often used objects like collections, maps etc. .
 *
 * @author daniel.kasmeroglu@kasisoft.net
 */
class Bucket<T>(val creator: Supplier<T>, val reset: Consumer<T>) {

    private val references = LinkedList<SoftReference<T>>()

    /**
     * Returns the number of references currently stored which means objects that can be acquired
     * without the need to create a new instance.
     *
     * @return   The number of references currently stored.
     */
    val size: Int
        get() = references.size

    val empty: Boolean
        get() = references.isEmpty()

    val notEmpty: Boolean
        get() = references.isNotEmpty()

    fun reset() {
        synchronized(references) {
            references.forEach { it.clear() }
            references.clear()
        }
    }

    /**
     * Creates a new object (potentially a reused one).
     *
     * @return   A new object.
     */
    fun allocate(): T {
        var result: T? = null
        synchronized(references) {
            while ((result == null) && references.isNotEmpty()) {
                val reference = references.removeAt(0)
                result = reference.get()
                reference.clear()
            }
        }
        return result ?: creator.get()
    }

    /**
     * Frees the supplied object, so it's allowed to be reused.
     *
     * @param object   The object that shall be freed.
     */
    fun free(obj: T?) {
        if (obj != null) {
            synchronized(references) {
                reset.accept(obj)
                references.add(SoftReference(obj))
            }
        }
    }

    /**
     * Executes the supplied function with the desired instance.
     *
     * @param function   The function that is supposed to be executed.
     *
     * @return   The return value of the supplied function.
     */
    fun <R> forInstance(function: Function<T, R>): R {
        val instance = allocate()
        return try {
            function.apply(instance)
        } finally {
            free(instance)
        }
    }

    /**
     * Executes the supplied consumer with the desired instance.
     *
     * @param consumer   The consumer that is supposed to be executed.
     */
    fun forInstanceDo(consumer: Consumer<T>) {
        val instance = allocate()
        try {
            consumer.accept(instance)
        } finally {
            free(instance)
        }
    }

} /* ENDCLASS */
